pandas/_libs/window/aggregations.pyx

```diff
@@ -1138,6 +1138,79 @@ def roll_quantile(const float64_t[:] values, ndarray[int64_t] start,
 
     return output
 
+def roll_rank(const float64_t[:] values, ndarray[int64_t] start,
+              ndarray[int64_t] end, int64_t minp, bint percentile) -> np.ndarray:
+    """
+    O(N log(window)) implementation using skip list
+
+    derived from roll_quantile
+    """
+    cdef:
+        Py_ssize_t i, j, s, e, N = len(values), idx
+        int rank = 0
+        int64_t nobs = 0, win
+        float64_t val
+        skiplist_t *skiplist
+        ndarray[float64_t] output
+
+    is_monotonic_increasing_bounds = is_monotonic_increasing_start_end_bounds(
+        start, end
+    )
+    # we use the Fixed/Variable Indexer here as the
+    # actual skiplist ops outweigh any window computation costs
+    output = np.empty(N, dtype=np.float64)
+
+    win = (end - start).max()
+    if win == 0:
+        output[:] = NaN
+        return output
+    skiplist = skiplist_init(<int>win)
+    if skiplist == NULL:
+        raise MemoryError("skiplist_init failed")
+
+    with nogil:
+        for i in range(0, N):
+            s = start[i]
+            e = end[i]
+
+            if i == 0 or not is_monotonic_increasing_bounds:
+                if not is_monotonic_increasing_bounds:
+                    nobs = 0
+                    skiplist = skiplist_init(<int>win)
```

Comment:
If we just reassign the pointer here, is it possible that we leak the previously pointed to skiplist?

Reply:
yeah this is a leak. I added a destroy call to get rid of it. This is also an issue in some of the other rolling functions. Should I fix those in this pr or make a separate issue for it?

Reply:
Best to make a separate issue since this will already be a pretty large pr (and smaller the pr, the easier to review)